src/main/scala/li/cil/oc/integration/forestry/UpgradeBeekeeper.scala

```diff
@@ -0,0 +1,197 @@
+  def swapDrone(context: Context, args: Arguments): Array[AnyRef] = {
+    val facing = checkSideForAction(args, 0)
+    withApiary(facing, housing => {
+      val selected = host.mainInventory.getStackInSlot(host.selectedSlot)
+      val oldQueen = housing.getBeeInventory.getDrone
+      housing.getBeeInventory.setDrone(selected)
+      host.mainInventory.setInventorySlotContents(host.selectedSlot, oldQueen)
+      result(true)
+    })
+  }
+
+  @Callback(doc = """function(side:number):number -- Get current progress percent for the apiary at the specified side.""")
+  def getBeeProgress(context: Context, args: Arguments): Array[AnyRef] = {
+    val facing = checkSideForAction(args, 0)
+    withApiary(facing, housing => result(housing.getBeekeepingLogic.getBeeProgressPercent))
+  }
+
+  @Callback(doc = """function(side:number):boolean -- Checks if current bee in the apiary at the specified side can work now.""")
+  def canWork(context: Context, args: Arguments): Array[AnyRef] = {
+    val facing = checkSideForAction(args, 0)
+    withApiary(facing, housing => result(housing.getBeekeepingLogic.canWork))
+  }
+
+  @Callback(doc = """function(honeyslot:number):boolean -- Analyzes bee in selected slot, uses honey from the specified slot.""")
+  def analyze(context: Context, args: Arguments): Array[AnyRef] = {
+    val inventory = host.mainInventory
+    val specimenSlot = host.selectedSlot
+    val specimen = inventory.getStackInSlot(specimenSlot)
+    if (!BeeManager.beeRoot.isMember(specimen))
+      return result(false, "Not a bee")
+
+    val honeySlot = args.checkSlot(inventory, 0)
+    val honeyStack = inventory.getStackInSlot(honeySlot)
+    if (honeyStack== null || honeyStack.stackSize == 0 || (honeyStack.getItem != PluginApiculture.items.honeydew && honeyStack.getItem != PluginApiculture.items.honeyDrop))
+      return result(false, "No honey!")
+
+    val individual = BeeManager.beeRoot.getMember(specimen)
+    if (!individual.isAnalyzed) {
+      individual.analyze
+      val nbttagcompound = new NBTTagCompound
+      individual.writeToNBT(nbttagcompound)
+      specimen.setTagCompound(nbttagcompound)
+      inventory.setInventorySlotContents(specimenSlot, specimen)
+      honeyStack.stackSize -= 1
+      inventory.setInventorySlotContents(honeySlot, honeyStack)
+    }
+    result(true)
+  }
+
+  private def findSameStack(upgrade: ItemStack, inv: ISidedInventory):Int = {
+    for (slot <- 2 to 5) {
+      val u = inv.getStackInSlot(slot)
+      if (u != null && u.getItem == upgrade.getItem && upgrade.getItemDamage == u.getItemDamage)
+        return slot
+    }
+    0
+  }
+
+  private def findEmptySlot(inv: ISidedInventory):Int = {
+    for (slot <- 2 to 5) {
+      if (inv.getStackInSlot(slot) == null)
+        return slot
+    }
+    0
+  }
+
+  @Callback(doc = """function(side:number):boolean -- Tries to add industrial upgrade from the selected slot to industrial apiary at the given side.""")
```

Comment:
It would be nice if these callbacks were only available if gendustry is installed. I believe this can be done using the `FilteredEnvironment` interface.